### ISSUE TITLE
SNI support

### ISSUE THREAD

**Issue by gsbasso** [forcedotcom/dataloader#106]
Now that Data Loader requires JRE1.8 I would have expected the Service Name Indicator (SNI) extension to be included in the SSL _Client Hello_ message, as this is now the default behaviour, In fact this is not the case.

Because I recently ran into a situation where the SNI extension was required I set out to see what would be needed to get Data Loader to send the SNI extension.

In the end I only needed to change the **getContent** method in **HttpClientTransport** (see code below). Note that I was not able to test the proxy settings.

I also updated the version of the Apache **httpclient** library dependency to 4.5.2 in the **pom.xml** file.

I would suggest including SNI support and updating the **httpclient** version in future releases

```
public InputStream getContent() throws IOException {
    //DefaultHttpClient client = new DefaultHttpClient();
    HttpClientBuilder clientBuilder = HttpClientBuilder.create();

    if (config.getProxy().address() != null) {
        String proxyUser = config.getProxyUsername() == null ? "" : config.getProxyUsername();
        String proxyPassword = config.getProxyPassword() == null ? "" : config.getProxyPassword();

        Credentials credentials;
        if (config.getNtlmDomain() != null && !config.getNtlmDomain().equals("")) {
            String computerName = InetAddress.getLocalHost().getCanonicalHostName();
            credentials = new NTCredentials(proxyUser, proxyPassword, computerName, config.getNtlmDomain());
        } else {
            credentials = new UsernamePasswordCredentials(proxyUser, proxyPassword);
        }

        InetSocketAddress proxyAddress = (InetSocketAddress) config.getProxy().address();
        HttpHost proxyHost = new HttpHost(proxyAddress.getHostName(), proxyAddress.getPort(), "http");
        //client.getParams().setParameter(ConnRoutePNames.DEFAULT_PROXY, proxyHost);
        clientBuilder.setProxy(proxyHost);

        AuthScope scope = new AuthScope(proxyAddress.getHostName(), proxyAddress.getPort(), null, null);
        //client.getCredentialsProvider().setCredentials(scope, credentials);
        CredentialsProvider provider = new BasicCredentialsProvider();
        provider.setCredentials(scope, credentials);
        clientBuilder.setDefaultCredentialsProvider(provider);
    }

    InputStream input = null;

    byte[] entityBytes = entityByteOut.toByteArray();
    HttpEntity entity = new ByteArrayEntity(entityBytes);
    post.setEntity(entity);

    try (CloseableHttpClient client = clientBuilder.build()) {

        // no changes in this section

    } finally {
        post.releaseConnection();
    }

    return input;
}
```

`

**Comment by rjmazzeo** [comment]
We were just made aware of the lack of SNI extensions from our support channel recently. It's on the radar now. I'll drop your method over ours and run the integration tests against it.

**Comment by rjmazzeo** [comment]
alternatively, you could fork and submit a pull request if you feel like it. Sounds like work though when you already have a solution in the comments.

**Comment by floppym** [comment]
This appears to have been resolved by b299db5d71a5a1093ea54423aaa867d1564c4bf5.

**Comment by ashitsalesforce** [comment]
closing the issue.
